muffnn/autoencoder/autoencoder.py

```diff
@@ -0,0 +1,682 @@
+"""
+Autoencoder in scikit-learn style with TensorFlow
+"""
+import logging
+import re
+import warnings
+
+import numpy as np
+import tensorflow as tf
+import scipy.sparse as sp
+
+from sklearn.base import TransformerMixin, BaseEstimator
+from sklearn.utils import check_array, check_random_state
+from sklearn.utils.validation import NotFittedError
+
+from ..core import TFPicklingBase, affine
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class Autoencoder(TFPicklingBase, TransformerMixin, BaseEstimator):
+    """
+    An autoencoder using TensorFlow.
+
+    Parameters
+    ----------
+    hidden_units : tuple or list, optional
+        A list of integers indicating the number of hidden layers and their
+        sizes. The list is reflected about the last element to form the
+        autoencoder. For example, if hidden_units is (256, 32, 16) then
+        the autoencoder has layers
+
+            [input_layer, 256, 32, 16, 32, 256, input_layer]
+
+    batch_size : int, optional
+        The batch size for learning and prediction. If there are fewer
+        examples than the batch size during fitting, then the the number of
+        examples will be used instead.
+    n_epochs : int, optional
+        The number of epochs (iterations through the training data) when
+        fitting.
+    dropout : float or None, optional
+        The dropout probability. If None, then dropout will not be used.
+        Note that dropout is not applied to the input layer.
+    activation : (callable, callable), optional
```

Comment:
related to the comment below about `metric`, i'd suggest having this argument just be for the hidden layer.

With respect to the output layer, I'd suggest either a) just not letting the user specifying the output activation for now, or b) having a separate kwarg that would allow specification of activations for discrete as well as ordinal variables.

I'd go for the (a) and maybe add that functionality later if it's really needed.

sigmoid/softmax seems more obviously correct for discrete variables and less likely to need changing than MSE for ordinal/continuous data, of course, but it seems strange to have some the activation set in stone for some types of data but not others.